### DISCUSSION TITLE
Using GCS emulator outside of `google-cloud-cpp`

### DISCUSSION THREAD

**Discussion by vnghia** [googleapis/google-cloud-cpp#6568] [Q&A] [answered]
Hi,

Currently, I am copying the code inside `google/cloud/storage/emulator` to use the emulator. I would like to ask:
- Is there a better way like: Dockerfile, pip package, etc ?
- I see that some other client library are using it too, how do they install it ?

**Comment by ghost** [comment, marked as answer]
I'm assuming you're talking about something like [this](https://github.com/googleapis/testing-infra-docker/tree/69e03f1c3789a28b9f11119011694c2b4a77b9c0/conformance-tests/storage/retry)?

Seems like they clone the repo, copy the emulator files over, install the python requirements then just run it.

**Reply by vnghia** [reply]
Thank @remyabel ! However, I got `permission-denied` for the Dockerfile. It seems that it is used internally only.

**Reply by ghost** [reply]
Permission denied where exactly? This is what I did:

```
cd google-cloud-cpp/
wget https://github.com/googleapis/testing-infra-docker/raw/69e03f1c3789a28b9f11119011694c2b4a77b9c0/conformance-tests/storage/retry/Dockerfile
docker build -f ./Dockerfile .
docker run 8a70b032a529
```

The only line I changed is the path (`google/cloud/storage/emulator`), but seems to work fine.

**Reply by vnghia** [reply]
Thank you ! I got `permission-denied` when I tried to pull from `docker run gcr.io/cloud-devrel-kokoro-resources/conformance-tests/storage/retry:${SHORT_SHA}_server-${CPP_SHORT_SHA}`. Do you know why ?

**Reply by ghost** [reply]
That looks like a private google registry.

**Reply by ghost** [reply]
Actually, that not may be the case. If it's supposed to be public, make sure your user is part of the docker group (or use sudo). docker login might also be necessary, probably not.

**Reply by coryan** [reply]
Right, the Docker image created from the `Dockerfile` is probably private and only intended for use by the library team.  You can still your own image using the same `Dockerfile`, and use it locally.  I am not aware of any reason why you could not publish this to docker hub or something similar.

**Reply by coryan** [reply]
Ah, I think @remyabel is right (as usual).

**Reply by vnghia** [reply]
Hmmm I still got `permission denied` after `gcloud auth configure-docker && docker login`. I think I will go with the copying method. Thank you !
